File: hermes-transport/maestro-transport.mjs
#!/usr/bin/env node
/**
 * Maestro Protocol — Hermes-Agent Transport (Node.js)
 * ====================================================
 * Gives a Hermes-Agent instance a native Maestro identity.
 *
 * Runs as a sidecar alongside the Hermes gateway. Listens for
 * inbound MaestroMessages, delivers them to Hermes via the API
 * server (/v1/chat/completions), and POSTs the reply back to
 * the original sender's /message endpoint.
 *
 * This makes Hermes-Agent a full Maestro peer — not a target
 * accessed via a proxy, but a first-class participant that can:
 *   - Receive direct messages from any Maestro agent
 *   - Participate in shared Blackboards
 *   - Be discovered via the shared registry file
 *   - Route replies back as MaestroMessages
 *
 * Usage:
 *   node maestro-transport.mjs [--config maestro_transport.json]
 *
 * Requirements:
 *   Node.js >= 18 (built-in fetch + http)
 *   No npm install needed.
 *
 * Config (maestro_transport.json):
 *   {
 *     "agentId": "hermes-lex",
 *     "port": 3844,
 *     "hermesApiUrl": "http://127.0.0.1:8642",
 *     "hermesApiKey": "maestro-local-dev",
 *     "conversation": "maestro",
 *     "registryPath": "/home/andjesse/.maestro/registry.json",
 *     "version": "3.2",
 *     "knownPeers": {
 *       "songbird": "http://10.0.2.2:3842/message"
 *     }
 *   }
 */

import http from 'http';
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// ─── Logging ──────────────────────────────────────────────────────────────────

function log(msg) {
  const ts = new Date().toTimeString().slice(0, 8);
  console.log(`[${ts}] [maestro] ${msg}`);
}

function logError(msg) {
  const ts = new Date().toTimeString().slice(0, 8);
  console.error(`[${ts}] [maestro] ERROR: ${msg}`);
}

// ─── Config ───────────────────────────────────────────────────────────────────

const DEFAULT_CONFIG = {
  agentId: 'hermes-lex',
  port: 3844,
  hermesApiUrl: 'http://127.0.0.1:8642',
  hermesApiKey: 'maestro-local-dev',
  conversation: 'maestro',
  registryPath: '.maestro/registry.json',
  version: '3.2',
  knownPeers: {},
};

function loadConfig(configPath) {
  // Resolve config path relative to this script's directory
  const resolved = path.isAbsolute(configPath)
    ? configPath
    : path.resolve(__dirname, configPath);

  if (fs.existsSync(resolved)) {
    try {
      const raw = fs.readFileSync(resolved, 'utf8');
      const loaded = JSON.parse(raw);
      return { ...DEFAULT_CONFIG, ...loaded };
    } catch (e) {
      logError(`Failed to parse config at ${resolved}: ${e.message}`);
      process.exit(1);
    }
  }
  log(`Config not found at ${resolved} — using defaults`);
  return { ...DEFAULT_CONFIG };
}

// ─── Registry (file-based, compatible with JS LocalRegistry) ──────────────────

class LocalRegistry {
  constructor(registryPath) {
    // Resolve relative paths from script directory
    this.path = path.isAbsolute(registryPath)
      ? registryPath
      : path.resolve(__dirname, registryPath);

    fs.mkdirSync(path.dirname(this.path), { recursive: true });
  }

  _load() {
    if (fs.existsSync(this.path)) {
      try {
        return JSON.parse(fs.readFileSync(this.path, 'utf8'));
      } catch {
        return [];
      }
    }
    return [];
  }

  _save(data) {
    fs.writeFileSync(this.path, JSON.stringify(data, null, 2), 'utf8');
  }

  register(agentId, webhookEndpoint, capabilities = []) {
    const data = this._load().filter(e => e.agentId !== agentId);
    data.push({
      agentId,
      webhookEndpoint,
      capabilities,
      registeredAt: Date.now(),
      lastSeen: Date.now(),
    });
    this._save(data);
    log(`Registered ${agentId} at ${webhookEndpoint}`);
  }

  lookup(agentId) {
    return this._load().find(e => e.agentId === agentId) ?? null;
  }

  unregister(agentId) {
    const data = this._load().filter(e => e.agentId !== agentId);
    this._save(data);
  }
}

// ─── Hermes Client ────────────────────────────────────────────────────────────

class HermesClient {
  constructor(apiUrl, apiKey) {
    this.apiUrl = apiUrl.replace(/\/$/, '');
    this.headers = {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${apiKey}`,
    };
  }

  _formatPrompt(message) {
    const lines = [
      '[Maestro Protocol — Inbound Message]',
      `From: ${message.sender?.agentId ?? 'unknown'}`,
      `Type: ${message.type ?? 'direct'}`,
    ];
    if (message.stageId) lines.push(`Connection: ${message.stageId}`);
    lines.push('');
    lines.push(message.content ?? '');
    return lines.join('\n');
  }

  async sendAndComplete(message) {
    const prompt = this._formatPrompt(message);

    const res = await fetch(`${this.apiUrl}/v1/chat/completions`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify({
        model: 'hermes-agent',
        messages: [{ role: 'user', content: prompt }],
        stream: false,
      }),
      signal: AbortSignal.timeout(120_000),
    });

    if (!res.ok) {
      const text = await res.text();
      logError(`Chat completions failed ${res.status}: ${text}`);
      return null;
    }

    const data = await res.json();
    const output = data?.choices?.[0]?.message?.content ?? null;
    if (output) log(`Response received: ${output.slice(0, 80)}...`);
    return output;
  }

  async healthCheck() {
    try {
      const res = await fetch(`${this.apiUrl}/health`, {
        signal: AbortSignal.timeout(5_000),
      });
      return res.ok;
    } catch {
      return false;
    }
  }
}

// ─── HTTP helpers ─────────────────────────────────────────────────────────────

function readBody(req) {
  return new Promise((resolve, reject) => {
    let body = '';
    req.on('data', chunk => (body += chunk));
    req.on('end', () => resolve(body));
    req.on('error', reject);
  });
}

function sendJSON(res, status, data) {
  const body = JSON.stringify(data);
  res.writeHead(status, {
    'Content-Type': 'application/json',
    'Content-Length': Buffer.byteLength(body),
  });
  res.end(body);
}

// ─── Transport ────────────────────────────────────────────────────────────────

class MaestroTransport {
  constructor(config) {
    this.config = config;
    this.agentId = config.agentId;
    this.port = config.port;
    this.hermes = new HermesClient(config.hermesApiUrl, config.hermesApiKey);
    this.registry = new LocalRegistry(config.registryPath);
    this.startedAt = null;
    this.server = http.createServer((req, res) => this._router(req, res));
  }

  async _router(req, res) {
    const url = new URL(req.url, `http://localhost`);

    if (req.method === 'GET' && url.pathname === '/health') {
      return this._handleHealth(req, res);
    }

    if (req.method === 'POST' &&
        (url.pathname === '/maestro/webhook' || url.pathname === '/message')) {
      return this._handleMessage(req, res);
    }

    if (req.method === 'GET' && url.pathname.startsWith('/connections/')) {
      const connectionId = url.pathname.split('/connections/')[1];
      return this._handleConnectionGet(req, res, connectionId);
    }

    sendJSON(res, 404, { error: 'Not found' });
  }

  _handleHealth(_req, res) {
    sendJSON(res, 200, {
      ok: true,
      agentId: this.agentId,
      platform: 'hermes-agent',
      uptime: this.startedAt ? Date.now() - this.startedAt : 0,
    });
  }

  async _handleMessage(req, res) {
    let message;
    try {
      const body = await readBody(req);
      message = JSON.parse(body);
    } catch {
      return sendJSON(res, 400, { accepted: false, reason: 'Invalid JSON' });
    }

    if (!message.id || !message.type || !message.sender) {
      return sendJSON(res, 400, { accepted: false, reason: 'Invalid message format' });
    }

    log(`Inbound message from ${message.sender?.agentId} type=${message.type}`);

    // Accept immediately, process async
    this._processMessage(message).catch(e => logError(`Unhandled error: ${e.message}`));
    sendJSON(res, 200, { accepted: true });
  }

  async _processMessage(message) {
    try {
      const output = await this.hermes.sendAndComplete(message);
      if (!output) {
        log('No output from Hermes — not routing reply');
        return;
      }

      const senderId = message.sender?.agentId;
      const senderReg = this.registry.lookup(senderId);
      if (!senderReg) {
        log(`Sender ${senderId} not in registry — cannot route reply`);
        return;
      }

      const reply = {
        id: crypto.randomUUID(),
        type: 'direct',
        content: output,
        sender: { agentId: this.agentId },
        recipient: senderId,
        timestamp: Date.now(),
        version: this.config.version,
        ...(message.stageId ? { stageId: message.stageId } : {}),
      };

      const endpoint = senderReg.webhookEndpoint;
      log(`Routing reply to ${senderId} at ${endpoint}: ${output.slice(0, 80)}...`);

      const replyRes = await fetch(endpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(reply),
        signal: AbortSignal.timeout(10_000),
      });

      if (replyRes.ok) {
        log(`Reply delivered to ${senderId}`);
      } else {
        logError(`Reply delivery failed: ${replyRes.status}`);
      }
    } catch (e) {
      logError(`Message processing error: ${e.message}`);
    }
  }

  _handleConnectionGet(_req, res, connectionId) {
    // Stub — full Connection support in a future version
    sendJSON(res, 200, {
      id: connectionId,
      status: 'active',
      agentId: this.agentId,
    });
  }

  async start() {
    // Health-check Hermes API before binding
    const ok = await this.hermes.healthCheck();
    if (!ok) {
      logError(`Hermes API not reachable at ${this.hermes.apiUrl} — is the gateway running?`);
      process.exit(1);
    }
    log(`Hermes API reachable at ${this.hermes.apiUrl}`);

    // Register self in local registry
    const endpoint = `http://0.0.0.0:${this.port}/maestro/webhook`;
    this.registry.register(this.agentId, endpoint);

    // Seed known peers (static cross-host endpoints)
    for (const [peerId, peerEndpoint] of Object.entries(this.config.knownPeers ?? {})) {
      this.registry.register(peerId, peerEndpoint);
      log(`Seeded known peer: ${peerId} at ${peerEndpoint}`);
    }

    this.startedAt = Date.now();

    await new Promise((resolve, reject) => {
      this.server.listen(this.port, '0.0.0.0', () => resolve());
      this.server.on('error', reject);
    });

    log(`${this.agentId} listening on 0.0.0.0:${this.port}`);
    log(`Hermes API: ${this.hermes.apiUrl}`);
    log('Ready.');
  }

  async stop() {
    this.registry.unregister(this.agentId);
    await new Promise(resolve => this.server.close(resolve));
    log('Stopped.');
  }
}

// ─── Entry point ──────────────────────────────────────────────────────────────

const args = process.argv.slice(2);
const configIdx = args.indexOf('--config');
const configPath = configIdx !== -1 ? args[configIdx + 1] : 'maestro_transport.json';

const config = loadConfig(configPath);
log(`Starting Maestro transport for ${config.agentId}`);

const transport = new MaestroTransport(config);
await transport.start();

log('Transport running. Press Ctrl+C to stop.');

process.on('SIGINT', async () => {
  log('SIGINT received — shutting down...');
  await transport.stop();
  process.exit(0);
});

process.on('SIGTERM', async () => {
  log('SIGTERM received — shutting down...');
  await transport.stop();
  process.exit(0);
});

// Keep alive
setInterval(() => {}, 1 << 30);

File: live/maestro_transport.py
#!/usr/bin/env python3
import asyncio
import json
import logging
import os
import re
import sys
import time
import uuid
from pathlib import Path
from typing import Any, Dict, Optional

try:
    from aiohttp import web, ClientSession, ClientTimeout
    AIOHTTP_AVAILABLE = True
except ImportError:
    AIOHTTP_AVAILABLE = False

logging.basicConfig(level=logging.INFO, format='[%(asctime)s] [maestro] %(message)s', datefmt='%H:%M:%S')
log = logging.getLogger(__name__)

DEFAULT_CONFIG = {
    "agentId": "hermes-lex", "port": 3844,
    "hermesApiUrl": "http://127.0.0.1:8642", "hermesApiKey": "maestro-local-dev",
    "conversation": "maestro", "registryPath": ".maestro/registry.json", "version": "3.2",
}

def load_config(path="maestro_transport.json"):
    if os.path.exists(path):
        with open(path) as f:
            return {**DEFAULT_CONFIG, **json.load(f)}
    return DEFAULT_CONFIG.copy()

import fcntl, threading

class LocalRegistry:
    def __init__(self, path):
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._lock = threading.Lock()
    def _load(self):
        if self.path.exists():
            try: return json.loads(self.path.read_text())
            except: pass
        return []
    def _save(self, data): self.path.write_text(json.dumps(data, indent=2))
    def register(self, agent_id, webhook_endpoint, capabilities=None):
        with self._lock:
            fd = None
            try:
                fd = os.open(str(self.path), os.O_RDWR | os.O_CREAT)
                fcntl.flock(fd, fcntl.LOCK_EX)
                size = os.lseek(fd, 0, os.SEEK_END)
                os.lseek(fd, 0, os.SEEK_SET)
                raw = os.read(fd, size).decode() if size else "[]"
                try:
                    data = json.loads(raw) if raw.strip() else []
                except Exception:
                    data = []
                data = [e for e in data if e.get("agentId") != agent_id]
                data.append({"agentId": agent_id, "webhookEndpoint": webhook_endpoint,
                    "capabilities": capabilities or [], "registeredAt": int(time.time()*1000), "lastSeen": int(time.time()*1000)})
                payload = json.dumps(data, indent=2)
                os.ftruncate(fd, 0)
                os.lseek(fd, 0, os.SEEK_SET)
                os.write(fd, payload.encode())
                log.info(f"Registered {agent_id} at {webhook_endpoint}")
            finally:
                if fd is not None:
                    fcntl.flock(fd, fcntl.LOCK_UN)
                    os.close(fd)
    def lookup(self, agent_id):
        with self._lock:
            for e in self._load():
                if e.get("agentId") == agent_id: return e
            return None
    def unregister(self, agent_id):
        with self._lock:
            fd = None
            try:
                fd = os.open(str(self.path), os.O_RDWR | os.O_CREAT)
                fcntl.flock(fd, fcntl.LOCK_EX)
                size = os.lseek(fd, 0, os.SEEK_END)
                os.lseek(fd, 0, os.SEEK_SET)
                raw = os.read(fd, size).decode() if size else "[]"
                try:
                    data = json.loads(raw) if raw.strip() else []
                except Exception:
                    data = []
                data = [e for e in data if e.get("agentId") != agent_id]
                payload = json.dumps(data, indent=2)
                os.ftruncate(fd, 0)
                os.lseek(fd, 0, os.SEEK_SET)
                os.write(fd, payload.encode())
                log.info(f"Unregistered {agent_id}")
            finally:
                if fd is not None:
                    fcntl.flock(fd, fcntl.LOCK_UN)
                    os.close(fd)

class HermesClient:
    def __init__(self, api_url, api_key, conversation):
        self.api_url = api_url.rstrip("/")
        self.api_key = api_key
        self.conversation = conversation
        self._headers = {"Content-Type": "application/json", "Authorization": f"Bearer {api_key}"}
    def _format_prompt(self, message):
        lines = ["[Maestro Protocol — Inbound Message]",
            f"From: {message.get('sender',{}).get('agentId','unknown')}",
            f"Type: {message.get('type','direct')}"]
        if message.get("stageId"): lines.append(f"Connection: {message['stageId']}")
        lines.extend(["", message.get("content", "")])
        return "\n".join(lines)
    async def send_and_complete(self, message):
        prompt = self._format_prompt(message)
        async with ClientSession() as session:
            async with session.post(f"{self.api_url}/v1/chat/completions",
                json={"model": "hermes-agent", "messages": [{"role": "user", "content": prompt}], "stream":
False},
                headers=self._headers, timeout=ClientTimeout(total=120)) as resp:
                if resp.status != 200:
                    log.error(f"Chat completions failed {resp.status}: {await resp.text()}")
                    return None
                data = await resp.json()
                output = data.get("choices",[{}])[0].get("message",{}).get("content")
                if output: log.info(f"Response: {output[:80]}...")
                return output
    async def health_check(self):
        try:
            async with ClientSession() as s:
                async with s.get(f"{self.api_url}/health", timeout=ClientTimeout(total=5)) as r:
                    return r.status == 200
        except: return False

class SeenSet:
    """TTL cache for deduplicating message IDs."""
    def __init__(self, ttl_seconds=300, max_size=5000):
        self._ttl = ttl_seconds * 1000
        self._max = max_size
        self._data = {}
    def add(self, msg_id):
        self._data[str(msg_id)] = int(time.time() * 1000)
        # simple prune
        if len(self._data) > self._max * 1.2:
            cutoff = int(time.time() * 1000) - self._ttl
            self._data = {k: v for k, v in self._data.items() if v > cutoff}
    def contains(self, msg_id):
        key = str(msg_id)
        if key in self._data:
            return True
        self.add(key)
        return False

class MaestroTransport:
    def __init__(self, config):
        self.config = config
        self.agent_id = config["agentId"]
        self.port = config["port"]
        self.hermes = HermesClient(config["hermesApiUrl"], config["hermesApiKey"], config["conversation"])
        self.registry = LocalRegistry(config["registryPath"])
        self.seen = SeenSet(ttl_seconds=300, max_size=5000)
        self._outbound_history = []  # (timestamp, recipient_id) for circuit breaker
        self._outbound_cooldowns = {}  # recipient_id -> cooldown_until_timestamp
        self._ack_pattern = re.compile(
            r'\b(Standing by|Copy|Acknowledged|Ready|On deck|Ack|Received|Got it|Roger|Will do|Understood|Noted)\b',
            re.IGNORECASE
        )
        self._loopback_count = 0
        self._received_ids = set()
        self.started_at = None
        self.app = web.Application()
        self.app.router.add_get("/health", self.handle_health)
        self.app.router.add_post("/message", self.handle_message)
        self.app.router.add_post("/maestro/webhook", self.handle_webhook)
        self.app.router.add_get("/connections/{connection_id}", self.handle_connection_get)
    async def handle_health(self, req):
        return web.json_response({"ok": True, "agentId": self.agent_id, "platform": "hermes-agent",
            "uptime": int(time.time()*1000) - self.started_at if self.started_at else 0,
            "seen_keys": len(self.seen._data)})
    async def handle_message(self, req):
        try: message = await req.json()
        except: return web.json_response({"accepted": False, "reason": "Invalid JSON"}, status=400)
        if not message.get("id") or not message.get("type") or not message.get("sender"):
            return web.json_response({"accepted": False, "reason": "Invalid message format"}, status=400)

        msg_id = message.get("id") or ""
        sender = message.get("sender",{}).get("agentId","unknown")

        # Prevent duplicate/broadcast loop
        if self.seen.contains(msg_id):
            self._loopback_count += 1
            if self._loopback_count % 10 == 1:
                log.info(f"Deduplicate {msg_id} from {sender} (total drops: {self._loopback_count})")
            return web.json_response({"accepted": True, "dedup": True})
        self.seen.add(msg_id)

        # Track this message ID as one we received
        self._received_ids.add(msg_id)
        # Also track replies based on our received messages
        if msg_id.startswith("reply-") or message.get("inReplyTo"):
            pass  # reply IDs are auto-generated, don't need special handling here

        # Structural: do not auto-reply to messages that are already replies (have inReplyTo)
        # Replies to replies cause echo loops
        if (message.get("inReplyTo") or message.get("reply_to")) and message.get("type") == "direct":
            log.info(f"Skipping auto-reply for {msg_id}: it is already a reply to another message")
            return web.json_response({"accepted": True, "no_reply": True})

        log.info(f"Inbound from {sender} type={message['type']}")
        asyncio.create_task(self._process_message(message))
        return web.json_response({"accepted": True})
    async def handle_webhook(self, req):
        """Compatibility shim: accept messages at /maestro/webhook too."""
        return await self.handle_message(req)
    async def _process_message(self, message):
        """Process inbound message, route reply to sender with loop protection."""
        try:
            output = await self.hermes.send_and_complete(message)
            if not output: log.warning("No output — not routing reply"); return
            content = output.strip() if isinstance(output, str) else str(output).strip() if output else ""
            # --- Outbound ACK filter ---
            if len(content) < 80 and self._ack_pattern.search(content):
                log.info(f"Outbound ACK filter dropped: {content[:60]}...")
                return
            sender_id = message["sender"]["agentId"]
            # --- Per-pair circuit breaker ---
            now_ts = time.time()
            self._outbound_history = [(t, r) for t, r in self._outbound_history if now_ts - t < 60]
            pair_count = sum(1 for t, r in self._outbound_history if r == sender_id)
            cooldown_until = self._outbound_cooldowns.get(sender_id, 0)
            if now_ts < cooldown_until:
                remaining = int(cooldown_until - now_ts)
                log.warning(f"Circuit breaker: cooling down on {sender_id} ({remaining}s left) — skipping reply")
                return
            if pair_count >= 3:
                self._outbound_cooldowns[sender_id] = now_ts + 120
                log.warning(f"Circuit breaker: tripped for {sender_id} (≥3 msgs/60s) — 2min cooldown")
                return
            self._outbound_history.append((now_ts, sender_id))
            sender_reg = self.registry.lookup(sender_id)
            if not sender_reg: log.warning(f"Sender {sender_id} not in registry"); return
            reply = {"id": str(uuid.uuid4()), "type": "direct", "content": output,
                "sender": {"agentId": self.agent_id}, "recipient": {"agentId": sender_id},
                "timestamp": int(time.time()*1000), "version": self.config["version"],
                "inReplyTo": message.get("id")}
            if message.get("stageId"): reply["stageId"] = message["stageId"]
            endpoint = sender_reg["webhookEndpoint"]
            log.info(f"Routing reply to {sender_id} at {endpoint}")
            async with ClientSession() as s:
                async with s.post(endpoint, json=reply, timeout=ClientTimeout(total=10)) as resp:
                    if resp.status == 200:
                        log.info(f"Reply delivered: {resp.status}")
                    else:
                        log.error(f"Reply failed: {resp.status}")
        except Exception as e: log.error(f"Processing error: {e}")
    async def handle_connection_get(self, req):
        return web.json_response({"id": req.match_info["connection_id"], "status": "active", "agentId":
self.agent_id})
    async def start(self):
        if not await self.hermes.health_check():
            log.error(f"Hermes API not reachable at {self.hermes.api_url}"); sys.exit(1)
        log.info(f"Hermes API reachable")
        # Seed knownPeers into registry so replies can route back
        for peer_id, peer_url in self.config.get("knownPeers", {}).items():
            self.registry.register(peer_id, peer_url)
            log.info(f"Seeded peer {peer_id} → {peer_url}")
        # Register self with 127.0.0.1 so other agents (Windows via VBox NAT + VM local) can reach us
        self.registry.register(self.agent_id, f"http://127.0.0.1:{self.port}/message")
        self.started_at = int(time.time()*1000)
        runner = web.AppRunner(self.app)
        await runner.setup()
        await web.TCPSite(runner, "0.0.0.0", self.port).start()
        log.info(f"{self.agent_id} listening on 0.0.0.0:{self.port}")
        log.info("Ready.")

async def main():
    config_path = "maestro_transport.json"
    if "--config" in sys.argv:
        idx = sys.argv.index("--config")
        if idx + 1 < len(sys.argv): config_path = sys.argv[idx + 1]
    if not AIOHTTP_AVAILABLE:
        print("ERROR: pip install aiohttp"); sys.exit(1)
    config = load_config(config_path)
    log.info(f"Starting Maestro transport for {config['agentId']}")
    transport = MaestroTransport(config)
    await transport.start()
    log.info("Transport running. Ctrl+C to stop.")
    try:
        while True: await asyncio.sleep(3600)
    except (KeyboardInterrupt, SystemExit):
        log.info("Shutting down.")

if __name__ == "__main__":
    asyncio.run(main())
